### PR TITLE
Feature/mds 2656 application and nda post

### PR DESCRIPTION
**Pull request by JustinMacaulay**
This is the first pass on a post to allow NROS and vFCBC to post now_submissions to our staging tables. Main files to look at:
- Static_Data.py
- Setup_marshmallow.py
- The application model in NoW_Submissions

### CODE REVIEW

File: services/core-api/app/api/now_submissions/resources/application_list_resource.py [hunk 1]
```diff
@@ -101,3 +106,27 @@ def _apply_filters_and_pagination(self,
             base_query = apply_sort(base_query, sort_criteria)
 
         return apply_pagination(base_query, page_number, page_size)
+
+    @api.doc(description='Save an application')
+    @requires_role_view_all
```
Comment (Jsyro) [inline services/core-api/app/api/now_submissions/resources/application_list_resource.py]: auth test?

File: services/core-api/app/api/now_submissions/resources/application_nda_resource.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+from app.extensions import api
+from app.api.now_submissions.models.application_nda import ApplicationNDA
+from app.api.now_submissions.response_models import APPLICATIONNDA
+from app.api.utils.access_decorators import requires_role_view_all
+from app.api.utils.resources_mixins import UserMixin 
+
+
+class ApplicationNDAResource(Resource, UserMixin ):
+    @api.doc(description='Fetch an application nda by id', params={})
+    @requires_role_view_all
```
Comment (Jsyro) [inline services/core-api/app/api/now_submissions/resources/application_nda_resource.py]: auth test?

File: services/core-api/app/api/utils/setup_marshmallow.py [hunk 3]
```diff
@@ -53,65 +58,69 @@ class CoreConverter(ModelConverter):
...
-                    class ModelSchema2(ModelSchema):
-                        activity_type_code = fields.String(dump_only=True)
-                        activity_detail_id = fields.Integer(dump_only=True)
-
-                    schema_class = type(schema_class_name, (ModelSchema2, ), {"Meta": Meta})
+                    schema_class = type(schema_class_name, (class_._ModelSchema, ), {"Meta": Meta})
 
                     setattr(class_, "_schema", schema_class)
-                    current_app.logger.debug(f'created {schema_class}')
+                    current_app.logger.debug(f'created {class_}')
```
Comment (Jsyro) [inline services/core-api/app/api/utils/setup_marshmallow.py]: can this be, created 'created schema for {class1}' or 'created {schema_class}'? just to be clear on what this does.

File: services/core-api/app/api/utils/setup_marshmallow.py [hunk 4]
```diff
@@ -53,65 +58,69 @@ class CoreConverter(ModelConverter):
...
-        #                     #exclude=[rel.backref.key] + [pk.name for pk in mapper.primary_keys])
-        #         except Exception as e:
-        #             raise e
+        for class_ in Base._decl_class_registry.values():
+            if hasattr(class_, "_schema"):
+                try:
+                    mapper = inspect(class_)
+                    for rel in mapper.relationships:
+                        if hasattr(rel.entity.class_, "_schema"):
+                            current_app.logger.debug(rel.key)
```
Comment (Jsyro) [inline services/core-api/app/api/utils/setup_marshmallow.py]: make this a better logger message? or no message at all?

File: services/core-api/app/api/utils/setup_marshmallow.py [hunk 5]
```diff
@@ -53,65 +58,69 @@ class CoreConverter(ModelConverter):
...
+                                rel.entity.class_._schema, many=rel.uselist)
+                            #exclude=[rel.backref.key] + [pk.name for pk in mapper.primary_keys])
+                except Exception as e:
+                    raise e
 
     return setup_schema_fn
 
 
 # TODO: finish this and resolve errors now_application/activity_detail_base.activity_type_code to all for programatic generation of schema
 # TODO: add call to model method to execute post_generation of schema.
```
Comment (Jsyro) [inline services/core-api/app/api/utils/setup_marshmallow.py]: You have actually solved this. please remove/update this comment.

File: services/core-api/app/api/utils/static_data.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+from sqlalchemy.dialects.postgresql import UUID
+
+from app.api.now_applications import models as app_models
+
+from app.api.constants import STATIC_DATA
+
+# This function is run right before setup_marshmallow and it looks through all of tables in our database.
+# It creates a mapping of classes to lists if their PK's, this is only done for code tables. To find the code tables
+# the classes are inspected and if they have a column named active_ind, it's type is not a UUID and its type is a string
+# the PK's are added to STATIC_DATA under the class name.
```
Comment (Jsyro) [inline services/core-api/app/api/utils/static_data.py]: Can you format this like a docstring """<>""" like in _deep_update_from_dict()? 

typo at 'lists **of** their PK's', don't need to list the conditions on what we are using to find a code, the code says that, maybe leave an inline comment that says 'code table conditions, to be refined later' or something like that.

File: services/core-api/app/api/utils/static_data.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+                                        active_ind=True).all()
+                            ]
+
+                    # This section is specific to NoW_submissions. Some of the code values that NROS and vFCBC send are
+                    # in long form so they are stored in the descriptions of the code tables so the descriptions of those
+                    # tables are also added under a (class name)_description in STATIC_DATA.
+
+                    if class_ in app_models.model_list and col.name == 'description':
+                        if type(pk.type) != UUID and pk.type.python_type == str:
+                            current_app.logger.debug(class_)
```
Comment (Jsyro) [inline services/core-api/app/api/utils/static_data.py]: debug message, improve or remove.

File: services/core-api/app/api/utils/static_data.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                            STATIC_DATA[f'{class_.__name__}_description'] = [
+                                a for a, in class_.query.with_entities(
+                                    getattr(class_, 'description', None)).filter_by(
+                                        active_ind=True).all()
+                            ]
+
+                    # if one of the now classes:
+                    #     add descriptions to static data.
+
+                    #exclude=[rel.backref.key] + [pk.name for pk in mapper.primary_keys])
```
Comment (Jsyro) [inline services/core-api/app/api/utils/static_data.py]: are these relevant?

File: services/core-api/app/api/utils/static_data.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            except Exception as e:
+                raise e
+
+    current_app.logger.debug(STATIC_DATA)
```
Comment (Jsyro) [inline services/core-api/app/api/utils/static_data.py]: improve message, maybe just list classes that static data includes? or the dictionary keys?

File: services/core-api/app/api/utils/setup_marshmallow.py (hunk 3 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/utils/setup_marshmallow.py]: Changed

File: services/core-api/app/api/now_submissions/resources/application_list_resource.py (hunk 1 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/now_submissions/resources/application_list_resource.py]: Done

File: services/core-api/app/api/utils/static_data.py (hunk 6 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/utils/static_data.py]: Changed to docstring style

File: services/core-api/app/api/utils/static_data.py (hunk 8 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/utils/static_data.py]: Removed

File: services/core-api/app/api/utils/static_data.py (hunk 7 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/utils/static_data.py]: removed

File: services/core-api/app/api/utils/static_data.py (hunk 9 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/utils/static_data.py]: removed

**Comment by Jsyro** [comment]
Inspect transmogrify for values that we read at 'Yes' or 'No' and validate those....

File: services/core-api/tests/now_submissions/resources/test_application_list_resource.py [hunk 10]
```diff
@@ -160,3 +642,47 @@ def test_get_now_application_list_filter_by_multiple_filters(self, test_client,
             str(submission.now_application_guid) not in map(lambda x: x['now_application_guid'],
                                                             get_data['records'])
             for submission in [now_submission_2, now_submission_4])
+
+    def test_post_now_application_happy_path(self, test_client, db_session, auth_headers):
```
Comment (exzizt) [inline services/core-api/tests/now_submissions/resources/test_application_list_resource.py]: happy path?

**Pull request merged**